manual/src/tutorials/memorymodel.etex

```diff
@@ -0,0 +1,783 @@
+answer is yes. With the C++ memory model, if the program has a data race, even
+in unrelated parts, then the semantics is undefined. If this snippet were
+linked with a library that had a data race, then, under the C++ memory model,
+the read may return any value. Since data races on unrelated locations can
+affect program behaviour, we say that C++ memory model is not bound in space.
+
+Unlike C++, Java memory model is bound in space. But Java memory model is not
+bound in time; data races in the future will affect the past behaviour. For
+example, consider the translation of this example to Java. We assume a prior
+definition of "Class c {int x;}" and a shared \emph{non-volatile} variable "C
+g". Now the snippet may be part of a larger program with parallel threads:
+
+\begin{verbatim}
+(* Thread 1 *)
+C c = new C();
+c.x = 42;
+a = c.x;
+g = c;
+
+(* Thread 2 *)
+g.x = 7;
+\end{verbatim}
+
+The read of "c.x" and the write of "g" in the first thread are done on separate
+memory locations. Hence, the Java memory model allows them to be reordered. As
+a result, the write in the second thread may occur before the read of "c.x",
+and hence, "c.x" returns "7".
+
+The OCaml equivalent of the Java code above is:
+
+\begin{caml_example*}{verbatim}
+let g = ref None
+
+let d1 () =
+  let c = ref 0 in
+  c := 42;
+  let a = !c in
+  g := Some c;
+  a
+
+let d2 () =
+  match !g with
+  | None -> ()
+  | Some c -> c := 7
+\end{caml_example*}
+
+Observe that there is a data race on both "g" and "c". Consider only the first
+three instructions in "d1":
+
+\begin{verbatim}
+let c = ref 0 in
+c := 42;
+let a = !c in
+...
+\end{verbatim}
+
+The OCaml memory model is bound both in space and time. The only memory location
+here is "c". Reasoning only about this snippet, there is neither the data race
+in space (the race on "g") nor in time (the future race on "c"). Hence, the
+snippet will have sequentially consistent behaviour, and the value returned by
+"!c" will be "42".
+
+\subsection{s:mm_msg_passing_again}{Message passing}
+
+Let us consider another example for reasoning with LDRF. Let us take a variant
+of the message passing example seen earlier (section \ref{s:mm_msg_passing}).
+
+\begin{caml_example*}{verbatim}
+let r = ref None
+let d1 () =
+  r := Some "Hello, world!"
+let d2 () =
+  match !r with
+  | None -> ()
+  | Some s -> print_endline s
+\end{caml_example*}
+
+The program uses a non-atomic reference "r" to share a message between domains.
+It is easy to see that there is a data race on "r". For the value "Some
+\"Hello, world!\"", OCaml allocates two objects -- one for the string and the
+other for the "Some" constructor. Both of these objects are first allocated
+uninitialised and then initialised with initialising writes. Are we guaranteed
+that when "d2" sees the "Some" variant when reading "r", the objects have been
+fully initialised?
+
+The answer is yes, and we can use LDRF to reason about this. Let us refactor the
```

Comment:
Thanks, it was interesting to see LDRF applied here. But I think it conflates OCaml's notion of initializing writes with the notion from the PLDI paper. I think the discussion around dependency ordering was convincing-enough to assume a stronger result of LDRF in the manual than in the paper, but you could say more clearly that you do so.

Reply:
Could you suggest the text that should go in here?